File: 2D Triangulated Mesh/Numerical Tests/Example1/NeuralField.py
import numpy as np

class NeuralField:
	"""
	NeuralField Object
	--------------------
	Discretized Amari type Neural Field 
	Integral equation, using a synaptic matrix
	assembled by the Nystrom Method.

	Receives:
		firing_rate -> object (as in FiringRate.py)
		synaptic_kernel -> object (as in Connectivity Kernel.py)
		mesh -> Mesh of interior points of each element
		n_elems -> Number of elements in the mesh
	"""
	def __init__(self, synaptic_matrix, firing_rate, ext_input):
		self.W = synaptic_matrix
		self.f_rate = firing_rate
		self.I = ext_input

	def evaluateAt(self, t, u, xs):
		"""
		Function: evaluateAt
		--------------------
		Evaluate the discretized integral equation
		at the given time t and vector u.
		""" 
		f_u = self.f_rate(u)
		I = np.zeros(len(xs))
		for i, x in enumerate(xs): 
			I[i] = self.I(t, x)
		return I - u + (self.W).dot(f_u)

File: C++/1D_NF_mmult/NeuralField.cpp
#include "NeuralField.hpp"

NeuralField::NeuralField(Matrix& synaptic_kernel,
                         FiringRate& firing_rate){
    _W = &synaptic_kernel;
    _f_rate = &firing_rate;
}

void NeuralField::evaluateAt(const double t,
                             const Vector& u,
                             Vector& new_u) const{

    // Evaluate firing rate function at vector u.
    Vector f_u(u.GetSize());
    _f_rate->evaluateAt(u, f_u);
    Vector S =  (*_W) * f_u;
    new_u = -u + S;
}

